frontend/catalyst/pennylane_extensions.py

```diff
@@ -2113,3 +2116,286 @@ def _callee():
+        for loc in batch_loc:
+            ax = in_axes_flat[loc]
+            fn_args_flat[loc] = jnp.take(args_flat[loc], 0, axis=ax)
+
+        fn_args = tree_unflatten(args_tree, fn_args_flat)
+
+        # Run 'fn' one time to get output-shape
+        init_result = fn(*fn_args, **kwargs)
+
+        # Check the validity of the output w.r.t. out_axes
+        out_axes_deep_struct = tree_structure(out_axes, is_leaf=lambda x: x is None)
+        init_result_deep_struct = tree_structure(init_result, is_leaf=lambda x: x is None)
+        if not isinstance(out_axes, int) and out_axes_deep_struct != init_result_deep_struct:
+            raise ValueError(
+                "Invalid 'out_axes'; it must be an int or match "
+                "the number of function results, but got "
+                f"{out_axes_deep_struct} axis specifiers and {init_result_deep_struct} results."
+            )
+
+        init_result_flat, init_result_tree = tree_flatten(init_result)
+
+        num_axes_out = len(init_result_flat)
+
+        if isinstance(out_axes, int):
+            out_axes_flat = [
+                out_axes,
+            ] * num_axes_out
+        else:
+            out_axes_flat, _ = tree_flatten(out_axes, is_leaf=lambda x: x is None)
+
+        out_loc = _get_batch_loc(out_axes_flat)
+
+        # Store batched results of all leaves
+        # in the flatten format with respect to the 'init_result' shape
+        batched_result_list = []
+        for j in range(num_axes_out):
+            out_shape = (
+                (batch_size,)
+                if not init_result_flat[j].shape
+                else (batch_size, *init_result_flat[j].shape)
+            )
+            batched_result_list.append(jnp.zeros(shape=out_shape, dtype=init_result_flat[j].dtype))
+            batched_result_list[j] = batched_result_list[j].at[0].set(init_result_flat[j])
+
+        # Apply mapping batched_args[1:] ---> fn(args)
+        @for_loop(1, batch_size, 1)
+        def loop_fn(i, batched_result_list):
+            fn_args_flat = args_flat
+            for loc in batch_loc:
+                ax = in_axes_flat[loc]
+                fn_args_flat[loc] = jnp.take(args_flat[loc], i, axis=ax)
+
+            fn_args = tree_unflatten(args_tree, fn_args_flat)
+            res = fn(*fn_args, **kwargs)
+
+            res_flat, _ = tree_flatten(res)
+
+            # Update the list of results
+            for j in range(num_axes_out):
+                batched_result_list[j] = batched_result_list[j].at[i].set(res_flat[j])
+
+            return batched_result_list
+
+        batched_result_list = loop_fn(batched_result_list)
+
+        # Support out_axes on dim > 0
+        for loc in out_loc:
+            if ax := out_axes_flat[loc]:
+                up_axes = [*range(batched_result_list[loc].ndim)]
+                up_axes[ax], up_axes[0] = up_axes[0], up_axes[ax]
+                batched_result_list[loc] = jnp.transpose(batched_result_list[loc], up_axes)
```

Comment:
Is this the second half of the procedure that we replaced above for inputs?

Reply:
No, this only applies to return values with `out_axes` values greater than 0. I found this implementation to be cleaner than `jnp.take` as here we want to re-shape the whole batch for those return values.

Reply:
Good point 👍